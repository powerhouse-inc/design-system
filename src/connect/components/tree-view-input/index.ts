export * from './tree-view-input';
