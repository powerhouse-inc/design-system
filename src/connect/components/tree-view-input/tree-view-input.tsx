import CheckIcon from '@/assets/icons/check-fill.svg';
import FolderIcon from '@/assets/icons/folder-close-fill.svg';
import CancelIcon from '@/assets/icons/xmark.svg';
import {
    TreeViewInput,
    TreeViewInputProps,
} from '@/powerhouse/components/tree-view-input';
import React from 'react';
import { twMerge } from 'tailwind-merge';
import { TreeItem } from '../tree-view-item';

export interface ConnectTreeViewInputProps
    extends Omit<
        TreeViewInputProps,
        | 'initialValue'
        | 'onSubmit'
        | 'onCancel'
        | 'submitIcon'
        | 'cancelIcon'
        | 'icon'
    > {
    item: TreeItem;
    onSubmit: (item: TreeItem) => void;
    onCancel: (item: TreeItem) => void;
}

export const ConnectTreeViewInput: React.FC<
    ConnectTreeViewInputProps
> = props => {
    const { className, item, onSubmit, onCancel, ...restProps } = props;

    return (
        <TreeViewInput
            icon={<img src={FolderIcon} />}
            submitIcon={<img src={CheckIcon} className="w-6 h-6" />}
            cancelIcon={
                <div className="w-6 h-6 flex items-center justify-center">
                    <img src={CancelIcon} alt="" />
                </div>
            }
            className={twMerge('bg-[#F1F5F9] rounded-lg h-12', className)}
            initialValue={item.label}
            onSubmit={value => onSubmit({ ...item, label: value })}
            onCancel={() => onCancel(item)}
            {...restProps}
        />
    );
};
