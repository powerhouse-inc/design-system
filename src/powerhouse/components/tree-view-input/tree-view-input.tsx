import React, { useState } from 'react';
import { useKeyboard } from 'react-aria';
import { Button, Input, PressEvent, TextField } from 'react-aria-components';
import ClickAwayListener from 'react-click-away-listener';
import { twMerge } from 'tailwind-merge';

export interface TreeViewInputProps
    extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onSubmit'> {
    icon?: React.ReactNode;
    cancelIcon?: React.ReactNode;
    submitIcon: React.ReactNode;
    level?: number;
    initialValue?: string;
    placeholder?: string;
    onCancel?: () => void;
    onSubmit: (value: string, event?: PressEvent) => void;
    'aria-label'?: string;
}

export const TreeViewInput: React.FC<TreeViewInputProps> = props => {
    const {
        icon,
        level = 0,
        submitIcon,
        cancelIcon,
        onSubmit,
        onCancel,
        className,
        style,
        placeholder,
        'aria-label': ariaLabel,
        initialValue = '',
        ...containerProps
    } = props;

    const [text, setText] = useState(initialValue);

    const { keyboardProps } = useKeyboard({
        onKeyUp(e) {
            if (e.key === 'Enter') {
                onSubmit(text);
            }
            if (e.key === 'Escape') {
                onCancel?.();
            }
        },
    });

    const paddingLeft = 10 * level + 24;

    return (
        <ClickAwayListener onClickAway={() => onSubmit(text)}>
            <div
                className={twMerge(
                    'flex flex-row items-center pr-5',
                    className,
                )}
                style={{
                    paddingLeft,
                    ...style,
                }}
                {...keyboardProps}
                {...containerProps}
            >
                {icon}
                <TextField
                    className="flex flex-1 ml-1 min-w-0"
                    value={text}
                    onChange={setText}
                    aria-label={ariaLabel}
                    autoFocus
                >
                    <Input
                        className="flex flex-1 outline-none bg-inherit min-w-0"
                        placeholder={placeholder}
                    />
                </TextField>
                <div className="flex flex-row items-center">
                    <Button
                        className="outline-none"
                        onPress={e => onSubmit(text, e)}
                    >
                        {submitIcon}
                    </Button>
                    {cancelIcon && (
                        <Button
                            onPress={onCancel}
                            className="outline-none ml-1"
                        >
                            {cancelIcon}
                        </Button>
                    )}
                </div>
            </div>
        </ClickAwayListener>
    );
};
