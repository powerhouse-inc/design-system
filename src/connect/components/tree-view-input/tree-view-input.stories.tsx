import { ActionType, ItemType } from '@/connect/types';
import type { Meta, StoryObj } from '@storybook/react';
import { ConnectTreeViewInput } from './tree-view-input';

const meta: Meta<typeof ConnectTreeViewInput> = {
    title: 'Connect/Components/TreeView',
    component: ConnectTreeViewInput,
    decorators: [
        Story => (
            <div className="bg-white p-10">
                <Story />
            </div>
        ),
    ],
    argTypes: {
        placeholder: { control: { type: 'text' } },
        level: { control: { type: 'number' } },
        onCancel: { control: { type: 'action' } },
        onSubmit: { control: { type: 'action' } },
        item: { control: { type: 'object' } },
    },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const TreeViewInput: Story = {
    args: {
        item: {
            id: 'MyDrive/Documents/MyDocuments',
            label: 'My-Documents',
            type: ItemType.Folder,
            action: ActionType.Update,
        },
        level: 0,
        placeholder: 'Folder Name',
    },
};
