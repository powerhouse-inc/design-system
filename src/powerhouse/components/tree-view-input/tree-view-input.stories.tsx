import type { Meta, StoryObj } from '@storybook/react';
import { Icon } from '..';
import { TreeViewInput } from './tree-view-input';

const icon = <Icon name="folder-open-fill" color="#6C7275" />;
const submitIcon = <Icon name="check-fill" color="#6C7275" />;
const cancelIcon = <Icon name="xmark" color="#6C7275" />;

const meta: Meta<typeof TreeViewInput> = {
    title: 'Powerhouse/Components/TreeView/TreeViewInput',
    component: TreeViewInput,
    argTypes: {
        icon: { control: { type: 'text' } },
        cancelIcon: { control: { type: 'text' } },
        submitIcon: { control: { type: 'text' } },
        placeholder: { control: { type: 'text' } },
        level: { control: { type: 'number' } },
        initialValue: { control: { type: 'text' } },
        onCancel: { control: { type: 'action' } },
        onSubmit: { control: { type: 'action' } },
    },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
    args: {
        icon,
        submitIcon,
        cancelIcon,
        initialValue: 'My Documents',
        level: 0,
        placeholder: 'Folder Name',
    },
};

export const WithStyles: Story = {
    decorators: [
        Story => (
            <div className="bg-white p-10">
                <Story />
            </div>
        ),
    ],
    args: {
        icon,
        submitIcon,
        cancelIcon,
        className: 'bg-[#F1F5F9] rounded-lg h-12',
        initialValue: 'My Documents',
        level: 0,
        placeholder: 'Folder Name',
    },
};
