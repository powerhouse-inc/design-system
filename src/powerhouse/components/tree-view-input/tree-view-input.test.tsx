import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { TreeViewInput } from './tree-view-input';

describe('TreeViewInput Component', () => {
    let onSubmit: jest.Mock<any, any, any>;
    let onCancel: jest.Mock<any, any, any>;

    const props = {
        level: 0,
        'aria-label': 'input',
        icon: 'Icon',
        initialValue: 'My Documents',
        submitIcon: 'Submit Icon',
        cancelIcon: 'Cancel Icon',
    };

    beforeEach(() => {
        // jest.useFakeTimers();
        onSubmit = jest.fn();
        onCancel = jest.fn();

        render(
            <TreeViewInput
                {...props}
                onSubmit={onSubmit}
                onCancel={onCancel}
            />,
        );
    });

    it('should match snapshot', () => {
        const { asFragment } = render(
            <TreeViewInput
                level={0}
                aria-label="input-label"
                onSubmit={() => {}}
                onCancel={() => {}}
                icon={<div>Icon</div>}
                initialValue="My Documents"
                submitIcon={<div>Submit Icon</div>}
                cancelIcon={<div>Cancel Icon</div>}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly', () => {
        expect(screen.getByText(props.icon)).toBeInTheDocument();
        expect(
            screen.getByDisplayValue(props.initialValue),
        ).toBeInTheDocument();
        expect(screen.getByText(props.submitIcon)).toBeInTheDocument();
        expect(screen.getByText(props.cancelIcon)).toBeInTheDocument();
    });

    it('should call onSubmit when click submit icon', () => {
        fireEvent.click(screen.getByText(props.submitIcon));

        expect(onSubmit).toHaveBeenCalled();
    });

    it('should call onSubmit when press enter key', () => {
        fireEvent.keyUp(screen.getByLabelText('input'), {
            key: 'Enter',
        });

        expect(onSubmit).toHaveBeenCalled();
    });

    it('should call onSubmit when click outside', async () => {
        await waitFor(() => screen.getByText(props.submitIcon), {
            timeout: 100,
        });
        fireEvent.click(document.body);

        expect(onSubmit).toHaveBeenCalled();
    });

    it('should call onCancel when click cancel icon', () => {
        fireEvent.click(screen.getByText(props.cancelIcon));

        expect(onCancel).toHaveBeenCalled();
    });

    it('should call onCancel when press Esc key', () => {
        fireEvent.keyUp(screen.getByLabelText('input'), {
            key: 'Escape',
        });

        expect(onCancel).toHaveBeenCalled();
    });

    it('should call onSubmit with text value', () => {
        fireEvent.change(screen.getByLabelText('input'), {
            target: { value: 'new value' },
        });

        fireEvent.click(screen.getByText(props.submitIcon));
        expect(onSubmit).toHaveBeenCalledWith('new value', expect.anything());
    });
});
