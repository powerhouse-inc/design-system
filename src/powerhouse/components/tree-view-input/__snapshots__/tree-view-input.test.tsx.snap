// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeViewInput Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row items-center pr-5"
    style="padding-left: 24px;"
  >
    <div>
      Icon
    </div>
    <div
      class="flex flex-1 ml-1 min-w-0"
      data-rac=""
    >
      <input
        aria-label="input-label"
        class="flex flex-1 outline-none bg-inherit min-w-0"
        data-focus-visible="true"
        data-focused="true"
        data-rac=""
        id="react-aria-:r5:"
        type="text"
        value="My Documents"
      />
    </div>
    <div
      class="flex flex-row items-center"
    >
      <button
        class="outline-none"
        data-rac=""
        type="button"
      >
        <div>
          Submit Icon
        </div>
      </button>
      <button
        class="outline-none ml-1"
        data-rac=""
        type="button"
      >
        <div>
          Cancel Icon
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
